src/app/core/export/query.service.ts

```diff
@@ -99,25 +119,41 @@ export class QueryService {
     }).value;
   }
 
-  private cacheRequiredData(query: string, from: Date, to: Date) {
+  private async cacheRequiredData(query: string, from: Date, to: Date) {
     const uncachedEntities = this.getUncachedEntities(query, from, to);
     const dataPromises = uncachedEntities.map((entity) => {
-      if (this.dataFunctions[entity.ENTITY_TYPE]) {
-        return this.dataFunctions[entity.ENTITY_TYPE](from, to).then(
-          (loadedEntities) => {
-            this.setEntities(entity, loadedEntities);
-            this.dataLoaded[entity.ENTITY_TYPE] = { from, to };
-          }
-        );
+      const info = this.entityInfo[entity.ENTITY_TYPE];
+      if (info?.dataFunction) {
+        return info.dataFunction(from, to).then((loadedEntities) => {
+          this.setEntities(entity, loadedEntities);
+          info.rangeLoaded = { from, to };
+        });
       } else {
         return this.entityMapper.loadType(entity).then((loadedEntities) => {
           this.setEntities(entity, loadedEntities);
-          this.dataLoaded[entity.ENTITY_TYPE] = true;
+          this.entityInfo[entity.ENTITY_TYPE] = { allLoaded: true };
         });
       }
     });
+    await Promise.all(dataPromises);
+    this.applyEntityUpdates(uncachedEntities);
+  }
 
-    return Promise.all(dataPromises);
+  private applyEntityUpdates(uncachedEntities: EntityConstructor[]) {
+    uncachedEntities
+      .filter(({ ENTITY_TYPE }) => !this.entityInfo[ENTITY_TYPE].updating)
+      .forEach(({ ENTITY_TYPE }) => {
+        this.entityInfo[ENTITY_TYPE].updating = true;
+        this.entityMapper
+          .receiveUpdates(ENTITY_TYPE)
+          .subscribe(({ entity, type }) => {
```

Comment:
will this not re-subscribe to the same entity type updates after a broader date range of that type is loaded?

And in general, when are we cleaning up these subscriptions? (or if we don't yet, shouldn't we clear that cache at some time rather than keeping all that stuff in memory forever after a user once used a report?)

Reply:
No, the `updating` flag is only set to `true` once and after that this entity type will always be filtered out here.

Generally, there is no concept of cleaning up yet as this service lives outside of the component. 
A option would be to just clear the cache once the user leaves the site but then this whole caching concept was pointless (as you can make changes without going away from the reporting page).